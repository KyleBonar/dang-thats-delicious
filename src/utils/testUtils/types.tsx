export interface MockLink {
  text: string;
  to: string;
  target: "_blank" | "_self";
}
