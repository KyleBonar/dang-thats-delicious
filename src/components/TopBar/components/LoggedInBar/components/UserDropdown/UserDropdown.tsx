import * as React from "react";

import styled from "../../../../../../theme/styled-components";
import Menu from "../Menu/Menu";
import Toggle from "../Toggle/Toggle";

const StyledDiv = styled.div`
  position: relative;
  margin-right: 2em;
`;
StyledDiv.displayName = "div";

export interface UserDropdownProps {}

const UserDropdown: React.FC<UserDropdownProps> = props => {
  const [isOpen, setIsOpen] = React.useState(false);

  const dropDownRef = React.useRef(null);

  // A memoized function for checking if the user's mouse click was inside of this component or not
  const onWindowClick = React.useCallback(
    event => {
      // Quick sanity check -- If the menu isn't even open, then we don't need to do anything
      if (!isOpen) return;

      // Find this component in the DOM
      const dropDownElem = dropDownRef.current as Element | null;
      if (dropDownElem === null) return;

      // Check for valid clicked target
      const isClickedElementValidNode = event.target instanceof Node;
      if (!isClickedElementValidNode) return;

      // Check if click was inside this FunctionalComponent or outside
      const isClickedElementWithinComponent = dropDownElem.contains(
        event.target
      );
      if (isClickedElementWithinComponent) return;

      setIsOpen(false); // Close dropdown
    },
    [isOpen, props.children]
  );

  // assign and remove window click event
  React.useEffect(() => {
    if (isOpen) {
      window.addEventListener("click", onWindowClick);
    } else {
      window.removeEventListener("click", onWindowClick);
    }

    return () => {
      // Make sure that listener is removed
      window.removeEventListener("click", onWindowClick);
    };
  }, [onWindowClick, isOpen]);

  return (
    <StyledDiv ref={dropDownRef}>
      <Toggle onClick={e => onToggleClick(e)} />

      {isOpen && <Menu />}
    </StyledDiv>
  );

  function onToggleClick(event?: React.MouseEvent): void {
    if (event) event.preventDefault();

    // Show or hide BodyChild based on state
    setIsOpen(!isOpen);
  }
};

export default UserDropdown;
