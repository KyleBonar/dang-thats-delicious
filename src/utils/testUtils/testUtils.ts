import * as React from "react";
import casual from "casual";
import { ReactWrapper } from "enzyme";
import configureStore from "redux-mock-store";

import { MockCard } from "./types";
import { AppState } from "../../redux/configureStore";
import { ISauce, ISaucesState } from "../../redux/sauces/types";
import { FlashMessageProps } from "../../components/FlashMessage/FlashMessage";
import Flatn from "../Flatn/Flatn";
import { IUserState } from "../../redux/users/types";

export const ITERATION_SIZE = 32;
const REACT_REGEX = /react(\d+)?./i;

// seed it so we get consistent results
casual.seed(777);

const relatedSauce = () => ({
  name: casual.name,
  slug: casual.uuid
});

export const fakeSauce = (): ISauce => ({
  _id: casual.random_element([undefined, casual.integer(1, 500)]),
  _addedToStore: casual.random_element([undefined, casual.unix_time]), // Unix time (in seconds) added to redux store
  _full: casual.random_element([undefined, casual.boolean]), // Whether we have full sauce or partial
  _related: casual.random_element([
    undefined,
    new Array(ITERATION_SIZE).fill(null).map(relatedSauce)
  ]), // List of related sauces
  isAdminApproved: casual.random_element([undefined, casual.boolean]),
  name: casual.name,
  ingredients: casual.string,
  author: casual.name,
  created: casual.unix_time, // Unix time (in seconds)
  types: casual.array_of_words(),
  maker: casual.name,
  description: casual.string,
  photo: casual.random_element([undefined, casual.url]),
  shu: casual.random_element([
    undefined,
    casual.integer(1, 500),
    casual.integer(1, 500).toString()
  ]),
  reviews: casual.random_element([
    undefined,
    new Array(ITERATION_SIZE).fill(null).map(x => casual.word)
  ]),
  city: casual.random_element([undefined, casual.city]),
  state: casual.random_element([undefined, casual.state]),
  country: casual.random_element([undefined, casual.country]),
  slug: casual.random_element([undefined, casual.uuid])
});

export const fakeSaucesState = (): ISaucesState => {
  const sauces: ISauce[] = new Array(casual.integer(1, 15))
    .fill(null)
    .map(fakeSauce);
  const { allSlugs, bySlug } = Flatn.saucesArr({ sauces });

  return {
    allSlugs,
    bySlug,
    total: allSlugs.length,
    query: undefined,
    saucesWithNewestReviews: undefined,
    newest: undefined,
    featured: undefined,
    types: casual.array_of_words(),
    orders: ["Newest", "Name", "Times Reviewed", "Avg Rating"]
  };
};

const fakeUsersState = (): IUserState => ({
  self: casual.random_element([
    undefined,
    {
      token: casual.random_element([undefined, casual.uuid]),
      displayName: casual.random_element([undefined, casual.name]),
      avatarURL: casual.random_element([undefined, casual.url]),
      isAdmin: casual.random_element([undefined, casual.boolean])
    }
  ]),
  byDisplayName: undefined,
  allDisplayNames: undefined
});

export const fakeStore = () => {
  const storeConfig = configureStore([]);

  const store: AppState = {
    sauces: fakeSaucesState(),
    users: fakeUsersState(),
    reviews: {
      byReviewID: undefined,
      allReviewIDs: undefined
    }
  };

  return storeConfig(store);
};

export const fakeJSXElement = (): JSX.Element => {
  return casual.random_element([
    React.createElement("div", { key: casual.uuid }, casual.string),
    React.createElement("span", { key: casual.uuid }, casual.string)
  ]);
};

const fakeCard = (): MockCard => ({
  title: casual.string,
  description: casual.string,
  to: casual.url,
  showLink: casual.boolean,
  imageLink: casual.url,
  className: casual.string,
  anchorText: casual.string
});

const wait = (amount = 0) =>
  new Promise(resolve => setTimeout(resolve, amount));

function simulateInputChange(
  wrapper: ReactWrapper,
  name: string,
  value: string
) {
  wrapper
    // .find(`input[name='${name}']`)
    .simulate("change", { target: { name, value } });
}

const isClassComponent = (component): boolean => {
  return (
    typeof component === "function" &&
    component.prototype &&
    !!component.prototype.isReactComponent
  );
};

// Ensure compatability with transformed code
const isFunctionComponent = (component): boolean => {
  return (
    typeof component === "function" &&
    String(component).includes("return") &&
    !!String(component).match(REACT_REGEX) &&
    String(component).includes(".createElement")
  );
};

const isElement = (typeElement): boolean => {
  return React.isValidElement(typeElement);
};

const isComponent = (component): boolean => {
  return isClassComponent(component) || isFunctionComponent(component);
};

const isDOMTypeElement = (typeElement): boolean => {
  return isElement(typeElement) && typeof typeElement.type === "string";
};

const generateValidPassword = (MIN_PASSWORD_LENGTH: number = 8) => {
  // generate length between minimum length and 3x the minimum length
  const _randomLength = casual.integer(
    MIN_PASSWORD_LENGTH,
    3 * MIN_PASSWORD_LENGTH
  );

  // turn generated length into random letters
  return new Array(_randomLength).fill(null).map(casual._letter).join("");
};

const generateInValidPassword = (MIN_PASSWORD_LENGTH: number = 8) => {
  // generate length between minimum length and 3x the minimum length
  const _randomLength = casual.integer(0, MIN_PASSWORD_LENGTH);

  // turn generated length into random letters
  return new Array(_randomLength).fill(null).map(casual._letter).join("");
};

export {
  casual,
  simulateInputChange,
  isDOMTypeElement,
  isComponent,
  wait,
  generateValidPassword
};
