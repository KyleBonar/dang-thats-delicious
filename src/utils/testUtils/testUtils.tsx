import * as React from "react";
import * as enzyme from "enzyme";
import casual from "casual";
import { ReactWrapper } from "enzyme";
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";

import { AppState } from "../../redux/configureStore";
import { ISauce, ISaucesState } from "../../redux/sauces/types";
import Flatn from "../Flatn/Flatn";
import { IUserState } from "../../redux/users/types";
import { IErrReturn } from "../Err/Err";
import { act } from "react-dom/test-utils";
import { Provider } from "react-redux";

export const ITERATION_SIZE = 32;
const REACT_REGEX = /react(\d+)?./i;

// seed it so we get consistent results
casual.seed(777);

const relatedSauce = () => ({
  name: casual.name,
  slug: casual.uuid
});

export const fakeSauce = (): ISauce => ({
  _id: casual.random_element([undefined, casual.integer(1, 500)]),
  _addedToStore: casual.random_element([undefined, casual.unix_time]), // Unix time (in seconds) added to redux store
  _full: casual.random_element([undefined, casual.boolean]), // Whether we have full sauce or partial
  _related: casual.random_element([
    undefined,
    new Array(ITERATION_SIZE).fill(null).map(relatedSauce)
  ]), // List of related sauces
  isAdminApproved: casual.random_element([undefined, casual.boolean]),
  name: casual.name,
  ingredients: casual.random_element([undefined, casual.string]),
  author: casual.name,
  created: casual.unix_time, // Unix time (in seconds)
  types: casual.random_element([undefined, casual.array_of_words()]),
  maker: casual.name,
  description: casual.string,
  photo: casual.random_element([undefined, casual.url]),
  shu: casual.random_element([
    undefined,
    casual.integer(1, 500),
    casual.integer(1, 500).toString()
  ]),
  reviews: casual.random_element([
    undefined,
    new Array(ITERATION_SIZE).fill(null).map(x => casual.word)
  ]),
  city: casual.random_element([undefined, casual.city]),
  state: casual.random_element([undefined, casual.state]),
  country: casual.random_element([undefined, casual.country]),
  slug: casual.random_element([undefined, casual.uuid])
});

export const generateFakeSauces = (): ISauce[] =>
  new Array(casual.integer(1, 15)).fill(null).map(fakeSauce);

export const fakeSaucesState = (): ISaucesState => {
  const sauces: ISauce[] = generateFakeSauces();
  const { allSlugs, bySlug } = Flatn.saucesArr({ sauces });

  const featured: undefined | string[] = casual.random_element([
    undefined,
    allSlugs.length > 0
      ? allSlugs.filter(slug => {
          if (casual.random < 0.4) return slug;
          return;
        })
      : undefined
  ]);

  return {
    allSlugs,
    bySlug,
    total: allSlugs.length,
    query: undefined,
    saucesWithNewestReviews: undefined,
    newest: undefined,
    featured,
    types: casual.array_of_words(),
    orders: ["Newest", "Name", "Times Reviewed", "Avg Rating"]
  };
};

const fakeUsersState = (): IUserState => ({
  self: casual.random_element([
    undefined,
    {
      token: casual.random_element([undefined, casual.uuid]),
      displayName: casual.random_element([undefined, casual.name]),
      avatarURL: casual.random_element([undefined, casual.url]),
      isAdmin: casual.random_element([undefined, casual.boolean])
    }
  ]),
  byDisplayName: undefined,
  allDisplayNames: undefined
});

export const fakeStore = () => {
  const middlewares = [thunk];
  const storeConfig = configureStore(middlewares);

  const store: AppState = {
    sauces: fakeSaucesState(),
    users: fakeUsersState(),
    reviews: {
      byReviewID: undefined,
      allReviewIDs: undefined
    }
  };

  return storeConfig(store);
};

export const fakeJSXElement = (): JSX.Element => {
  return casual.random_element([
    React.createElement("div", { key: casual.uuid }, casual.string),
    React.createElement("span", { key: casual.uuid }, casual.string)
  ]);
};

const wait = (amount = 0) =>
  new Promise(resolve => setTimeout(resolve, amount));

function simulateInputChange(
  wrapper: ReactWrapper,
  name: string,
  value: string
) {
  wrapper
    // .find(`input[name='${name}']`)
    .simulate("change", { target: { name, value } });
}

const isClassComponent = (component): boolean => {
  return (
    typeof component === "function" &&
    component.prototype &&
    !!component.prototype.isReactComponent
  );
};

// Ensure compatability with transformed code
const isFunctionComponent = (component): boolean => {
  return (
    typeof component === "function" &&
    String(component).includes("return") &&
    !!String(component).match(REACT_REGEX) &&
    String(component).includes(".createElement")
  );
};

const isElement = (typeElement): boolean => {
  return React.isValidElement(typeElement);
};

const isComponent = (component): boolean => {
  return isClassComponent(component) || isFunctionComponent(component);
};

const isDOMTypeElement = (typeElement): boolean => {
  return isElement(typeElement) && typeof typeElement.type === "string";
};

const generateValidPassword = (MIN_PASSWORD_LENGTH: number = 8): string => {
  // generate length between minimum length and 3x the minimum length
  const _randomLength = casual.integer(
    MIN_PASSWORD_LENGTH,
    3 * MIN_PASSWORD_LENGTH
  );

  // turn generated length into random letters
  return new Array(_randomLength).fill(null).map(casual._letter).join("");
};

const generateInValidPassword = (MIN_PASSWORD_LENGTH: number = 8): string => {
  // generate length between minimum length and 3x the minimum length
  const _randomLength = casual.integer(1, MIN_PASSWORD_LENGTH - 1);

  // turn generated length into random letters
  return new Array(_randomLength).fill(null).map(casual._letter).join("");
};

// Function designed to resemble how a network error would look like.
// This will help the catch(err) only have to handle a single format
const generateErr = (): IErrReturn => {
  return {
    response: {
      data: {
        status: 400,
        msg: casual.string,
        isGood: false
      }
    }
  };
};

// lets us mount up a simple custom hook
const mountReactHook = hook => {
  const Component = ({ children }) => children(hook());
  const componentHook = {};
  let componentMount;

  act(() => {
    componentMount = enzyme.shallow(
      <Component>
        {hookValues => {
          Object.assign(componentHook, hookValues);
          return null;
        }}
      </Component>
    );
  });
  return { componentMount, componentHook };
};

export interface ImountReactHookWithReduxStore {
  componentMount: any;
  componentHook: any;
}
const mountReactHookWithReduxStore = (
  hook,
  store
): ImountReactHookWithReduxStore => {
  const Component = ({ children }) => children(hook());
  const componentHook = {};
  let componentMount;

  act(() => {
    componentMount = enzyme.mount(
      <Provider store={store}>
        <Component>
          {hookValues => {
            Object.assign(componentHook, hookValues);
            return null;
          }}
        </Component>
      </Provider>
    );
  });
  return { componentMount, componentHook };
};

export {
  casual,
  simulateInputChange,
  isDOMTypeElement,
  isComponent,
  wait,
  generateValidPassword,
  generateInValidPassword,
  generateErr,
  mountReactHook,
  mountReactHookWithReduxStore
};
