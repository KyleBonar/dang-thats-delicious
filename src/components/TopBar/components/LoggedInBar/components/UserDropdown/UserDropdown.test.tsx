import * as React from "react";
import * as enzyme from "enzyme";

import UserDropdown from "./UserDropdown";
import { ITERATION_SIZE } from "../../../../../../utils/testUtils/testUtils";

describe("<UserDropdown />", () => {
  let wrappers: any = [];

  beforeAll(() => {
    wrappers = new Array(ITERATION_SIZE).fill(null).map(() => {
      return enzyme.shallow(<UserDropdown />);
    });
  });

  afterEach(() => {
    // Close Menu if open
    wrappers.forEach(wrapper => {
      if (wrapper.find("Menu").exists()) {
        wrapper.find("Toggle").simulate("click");
      }
    });
  });

  it("renders", () => {
    wrappers.forEach(wrapper => {
      expect(wrapper.exists()).toBeTruthy();
    });
  });

  it("matches snapshot", () => {
    wrappers.forEach(wrapper => {
      expect(wrapper.exists()).toMatchSnapshot();
    });
  });

  it("renders Toggle component", () => {
    wrappers.forEach(wrapper => {
      expect(wrapper.find("Toggle").exists()).toBeTruthy();
    });
  });

  it("renders Toggle component", () => {
    wrappers.forEach(wrapper => {
      expect(wrapper.find("Toggle").exists()).toBeTruthy();
    });
  });

  it("does not render Menu component initially", () => {
    wrappers.forEach(wrapper => {
      expect(wrapper.find("Menu").exists()).toBeFalsy();
    });
  });

  it("renders Menu component after Toggle is clicked", () => {
    wrappers.forEach(wrapper => {
      wrapper.find("Toggle").simulate("click");

      expect(wrapper.find("Menu").exists()).toBeTruthy();
    });
  });

  it("closes Menu component after Toggle is clicked twice", () => {
    wrappers.forEach(wrapper => {
      wrapper.find("Toggle").simulate("click");
      wrapper.find("Toggle").simulate("click");

      expect(wrapper.find("Menu").exists()).toBeFalsy();
    });
  });
});
