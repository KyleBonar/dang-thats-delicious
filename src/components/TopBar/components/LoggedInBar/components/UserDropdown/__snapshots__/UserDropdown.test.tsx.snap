// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UserDropdown /> matches snapshot 1`] = `true`;

exports[`<UserDropdown /> matches snapshot 2`] = `true`;

exports[`<UserDropdown /> matches snapshot 3`] = `true`;

exports[`<UserDropdown /> matches snapshot 4`] = `true`;

exports[`<UserDropdown /> matches snapshot 5`] = `true`;

exports[`<UserDropdown /> matches snapshot 6`] = `true`;

exports[`<UserDropdown /> matches snapshot 7`] = `true`;

exports[`<UserDropdown /> matches snapshot 8`] = `true`;

exports[`<UserDropdown /> matches snapshot 9`] = `true`;

exports[`<UserDropdown /> matches snapshot 10`] = `true`;

exports[`<UserDropdown /> matches snapshot 11`] = `true`;

exports[`<UserDropdown /> matches snapshot 12`] = `true`;

exports[`<UserDropdown /> matches snapshot 13`] = `true`;

exports[`<UserDropdown /> matches snapshot 14`] = `true`;

exports[`<UserDropdown /> matches snapshot 15`] = `true`;

exports[`<UserDropdown /> matches snapshot 16`] = `true`;

exports[`<UserDropdown /> matches snapshot 17`] = `true`;

exports[`<UserDropdown /> matches snapshot 18`] = `true`;

exports[`<UserDropdown /> matches snapshot 19`] = `true`;

exports[`<UserDropdown /> matches snapshot 20`] = `true`;

exports[`<UserDropdown /> matches snapshot 21`] = `true`;

exports[`<UserDropdown /> matches snapshot 22`] = `true`;

exports[`<UserDropdown /> matches snapshot 23`] = `true`;

exports[`<UserDropdown /> matches snapshot 24`] = `true`;

exports[`<UserDropdown /> matches snapshot 25`] = `true`;

exports[`<UserDropdown /> matches snapshot 26`] = `true`;

exports[`<UserDropdown /> matches snapshot 27`] = `true`;

exports[`<UserDropdown /> matches snapshot 28`] = `true`;

exports[`<UserDropdown /> matches snapshot 29`] = `true`;

exports[`<UserDropdown /> matches snapshot 30`] = `true`;

exports[`<UserDropdown /> matches snapshot 31`] = `true`;

exports[`<UserDropdown /> matches snapshot 32`] = `true`;
